File: _ETC/B형대비문제/Social media/Social media/정답코드/Java/fast/1.java
//59,604KB	247MS	JAVA	2,585B
class UserSolution {
    private static final int USER_MAX = 1001;
    private static final int POST_MAX = 100001;
    private static final int FEED_MAX = 10;
     
    private static byte tc = 0;
    private static int postNu;
     
    //private static final int[] feedPost = new int[FEED_MAX + 1];
    private static final int[] postPoster = new int[POST_MAX];
    private static final int[] postID = new int[POST_MAX];
    private static final int[] postLike = new int[POST_MAX];
    private static final int[] postTime = new int[POST_MAX];
    private static final int[] idPost = new int[POST_MAX];
    private static final byte[][] fol = new byte[USER_MAX][USER_MAX];
     
    public void init(int N)
    {
        int i, j;
        tc++;
        if (tc <= 1) {
            for (i = USER_MAX-1; i > 0; i--) {
                for (j = USER_MAX-1; j > 0; j--) {
                    fol[i][j] = 0;
                }
            }
        }
        for (i = N; i > 0; i--) {
            fol[i][i] = tc;
        }
        postNu = 0;
    }
 
    public void follow(int uID1, int uID2, int timestamp)
    {
        fol[uID1][uID2] = tc;
    }
 
    public void makePost(int uID, int pID, int timestamp)
    {
        postID[postNu] = pID;
        idPost[pID] = postNu;
        postPoster[postNu] = uID;
        postLike[postNu] = 0;
        postTime[postNu] = timestamp;
        postNu++;
    }
 
    public void like(int pID, int timestamp)
    {
        postLike[idPost[pID]]++;
    }
 
    public void getFeed(int uID, int timestamp, int pIDList[])
    {
        int fNu, p, l, r;
        final int[] feedPost = new int[FEED_MAX + 1];
        final boolean[] feedTill = new boolean[FEED_MAX + 1];
        boolean stop, till1000;
        fNu = 0;
        p = postNu;
        while (p > 0) {
            p--;
            if (fol[uID][postPoster[p]] == tc) {
                till1000 = (timestamp - postTime[p] <= 1000);
                r = fNu;
                l = r-1;
                while (l >= 0) {
                    if (till1000) {
                        if (feedTill[l]) {
                            if (postLike[p] > postLike[feedPost[l]]) {
                                stop = false;
                            } else if (postLike[p] < postLike[feedPost[l]]) {
                                stop = true;
                            } else {
                                if (postTime[p] > postTime[feedPost[l]]) {
                                    stop = false;
                                } else {
                                    stop = true;
                                }
                            }
                        } else {
                            stop = false;
                        }                       
                    } else {
                        if (feedTill[l]) {
                            stop = true;
                        } else {
                            if (postTime[p] > postTime[feedPost[l]]) {
                                stop = false;
                            } else {
                                stop = true;
                            }
                        }
                    }
                    if (stop) break;
                    feedPost[r] = feedPost[l];
                    feedTill[r] = feedTill[l];
                    r--;
                    l--;
                }
                feedPost[r] = p;
                feedTill[r] = till1000;
                if (fNu < 10) {
                    fNu++;
                } else {
                    if (!till1000) break;
                }
            }
        }
        l = fNu;
        while (l > 0) {
            l--;
            pIDList[l] = postID[feedPost[l]];
        }
    }
}

File: _ETC/B형대비문제/Social media/Social media/정답코드/CPP/fast/1.cpp
//22,776KB	97MS	C++	2,217B	
#define fi(i,a,b) for(register int i=a; i<b; ++i)
#define fde(i,a,b) for(register int i=a; i>=b; --i)
 
int post_uid[101000];
int post_likes[101000];
int post_ts[101000];
int post_next[101000];
 
int last[1002], old[1002];
int pc;
int n_;
int fw[1002][1002];
int fwc[1002];
 
void init(int N)
{
    n_ = N;
    pc = 0;
    fi(i, 1, n_ + 1) fwc[i] = 1, fw[i][0] = i, last[i] = 0;
}
 
void follow(int uID1, int uID2, int timestamp)
{
    fw[uID1][fwc[uID1]++] = uID2;
}
 
void makePost(int uID, int pID, int timestamp)
{
    post_likes[pID] = 0;
    post_uid[pID] = uID;
    post_ts[pID] = timestamp;
    post_next[pID] = last[uID];
    last[uID] = pID;
    pc = pID;
}
 
void like(int pID, int timestamp)
{
    post_likes[pID]++;
}
 
#define LESS(a,b) a.val < b.val
template<class T>
void qsrt(T* v, int left, int right)
{
    if (left >= right || left >= 10) return;
    register T t;
    if (left + 1 == right) { if (LESS(v[right], v[left])) { t = v[right]; v[right] = v[left]; v[left] = t; } return; }
    register T pivot = v[left + (right - left) / 2];
 
    register int i = left, j = right;
    while (true) {
        while (LESS(v[i], pivot)) i++;
        while (LESS(pivot, v[j])) j--;
        if (i >= j) break;
        t = v[i]; v[i] = v[j]; v[j] = t;
        i++; j--;
    }
    qsrt(v, left, j);
    qsrt(v, j + 1, right);
}
 
struct { int val, idx; } d[100000];
 
void getFeed(int uID, int timestamp, int res[])
{
    register int aidx = 0;
    register int u;
    timestamp -= 1000;
    fi(i, 0, fwc[uID]) {
        u = fw[uID][i];
        register int p = last[u];
        while (p && post_ts[p] >= timestamp) {
            d[aidx].val = -(post_likes[p] * 101000 + post_ts[p]);
            d[aidx].idx = p;
            ++aidx;
            p = post_next[p];
        }
        old[u] = p;
    }
    register int need = 10 - aidx;
    if (need >0)
        fi(i, 0, fwc[uID]) {
        u = fw[uID][i];
        register int p = old[u];
        fi(j, 0, need) {
            if (!p) break;
            d[aidx].val = -post_ts[p];
            d[aidx].idx = p;
            ++aidx;
            p = post_next[p];
        }
    }
    qsrt(d, 0, aidx - 1);
    if (aidx > 10) aidx = 10;
    fi(i, 0, aidx) res[i] = d[i].idx;
}